### PR TITLE
Add support for `Accessor.buffer`

### PR DESCRIPTION
**Pull request by ibgreen**
# For #702

#### Background
- Allow accessor objects to reference a `buffer`, this greatly supports the interleaved use case that is common in glTF models, and currently weakly supported by luma.

Multiple accessor referencing the same buffer:
```
const buffer = // "interleaved" vertex attributes: 3 floats for position followed by 4 bytes for RGBA
model.setAttributes({
  positions: {buffer, stride: 16, offset: 0, ...}}),
  colors: {buffer, stride: 16, offset: 12, ...}})
}
```

#### Change List
- Accessor changes
- VertexArray accepts accessors with buffer
- buffer code cleanup

### CODE REVIEW

File: modules/core/src/core/attribute.js [hunk 1]
```diff
@@ -122,11 +122,11 @@ export default class Attribute {
     this.divisor = divisor;
 
     if (isInstanced !== undefined) {
-      log.deprecated('Attribute.isInstanced');
+      log.deprecated('Attribute.isInstanced')();
       this.divisor = isInstanced ? 1 : 0;
     }
     if (instanced !== undefined) {
-      log.deprecated('Attribute.instanced');
+      log.deprecated('Attribute.instanced')();
```
Comment (georgios-uber) [inline modules/core/src/core/attribute.js]: you need a linter rule for this!

File: dev-docs/RFCs/README.md [hunk 2]
```diff
@@ -32,6 +32,7 @@ These are early ideas not yet associated with any release
 | --- | ---    | ---    | ---         |
 | [**luma.gl Module Structure**](v6.2/module-structure-rfc.md) | @ibgreen | **Partially Implemented** |  Outlines module structure to be used in monorepo transition. Implementeation continues from v6.2. |
 | [**glTF2 support**](v7.0/gltf-support-rfc.md) | @ibgreen | **Draft** | Proposes a multi-step approach to full glTF2 support. |
+| [**Attribute/Accessor Class Alignment**](v7.0/attribute-accessor-alignment-rfc.md) | @ibgreen | **Draft** | Ensure we have a consistent logical API for specifying attribute and buffer accessor metadata. |
```
Comment (georgios-uber) [inline dev-docs/RFCs/README.md]: This file does not exist!

File: modules/core/src/webgl/buffer.js [hunk 3]
```diff
@@ -2,23 +2,55 @@ import GL from '@luma.gl/constants';
...
-import {getGLTypeFromTypedArray, getTypedArrayFromGLType} from '../webgl-utils/typed-array-utils';
-import {log} from '../utils';
-import assert from '../utils/assert';
+import {getGLTypeFromTypedArray, getTypedArrayFromGLType} from '../webgl-utils';
+import {log, assert, checkProps} from '../utils';
 
 const DEBUG_DATA_LENGTH = 10;
 
+// Shared prop checks for constructor and setProps
+const DEPRECATED_PROPS = {
```
Comment (georgios-uber) [inline modules/core/src/webgl/buffer.js]: So we are getting rid of the accessor?

File: modules/core/src/webgl/vertex-array.js [hunk 4]
```diff
@@ -318,6 +304,30 @@ export default class VertexArray {
...
+      const accessor = value[1];
+      this.setBuffer(locationOrName, buffer, accessor);
+    } else if (ArrayBuffer.isView(value) || Array.isArray(value)) {
+      // Signature: {attributeName: constant}, constant == short (typed) array
+      const constant = value;
+      this.setConstant(locationOrName, constant);
+    } else if (value.buffer instanceof Buffer) {
+      // luma.gl v7: Support accessor objects with 'buffer' field
+      // for interleaved data
+      // Signature: {attributeName: {...accessor, buffer}}
```
Comment (georgios-uber) [inline modules/core/src/webgl/vertex-array.js]: 👍 I like the signature comments but please make them
new Buffer(gl, ......)

File: test/browser.js [hunk 5]
```diff
@@ -27,7 +27,7 @@ test.onFinish(() => callExposedFunction('testDone', {success: true}));
 test.onFailure(() => callExposedFunction('testDone', {success: false}));
 
 test('Browser tests', t => {
-  require('./index-webgl-independent-tests');
+  // require('./index-webgl-independent-tests');
```
Comment (georgios-uber) [inline test/browser.js]: why?

File: test/index-webgl-dependent-tests.js [hunk 6]
```diff
@@ -9,5 +9,4 @@ import './modules/core/webgl-context';
 import './modules/core/webgl';
 import './modules/core/core';
 
-// GPGPU is a set of shadertool modules, no WebGL dependencies
```
Comment (georgios-uber) [inline test/index-webgl-dependent-tests.js]: Why?

File: modules/core/src/core/attribute.js (hunk 1 above)
Comment (ibgreen) [reply modules/core/src/core/attribute.js]: > you need a linter rule for this!

Yes, that is a good idea! Do you know if it possible to define custom linter rules for our own APIs? Any link to how to do it?

File: modules/core/src/webgl/buffer.js (hunk 3 above)
Comment (ibgreen) [reply modules/core/src/webgl/buffer.js]: On the contrary, I am enforcing the accessor in favor of allowing accessor fields directly in the Buffer props. 

What these rules are saying is that:

`new Buffer(gl, {size: ...})` should be replaced with `new Buffer(gl, {accessor {size: ...}})`

File: modules/core/src/webgl/vertex-array.js (hunk 4 above)
Comment (ibgreen) [reply modules/core/src/webgl/vertex-array.js]: These are not parameters to `Buffer` but the values of each entry in the attributes map.

Are you suggesting:

```
Signature: {attributeName: {...accessor, buffer: new Buffer(gl, ...)}}
```

I think the key is updating this in a good way in the docs, not here. Probably more should be done in that regard.

File: test/index-webgl-dependent-tests.js (hunk 6 above)
Comment (ibgreen) [reply test/index-webgl-dependent-tests.js]: Turns out 1chandu's GPGPU module is not WebGL-independent so I corrected that statement.

**Comment by ibgreen** [comment]
Will add some tests before I land but would be good to get the changes approved.

File: modules/core/src/core/attribute.js (hunk 1 above)
Comment (georgios-uber) [reply modules/core/src/core/attribute.js]: https://github.com/uber/luma.gl/commit/ef34ca48db2fceb42714fe8891cf8951440f1041

**Comment by ibgreen** [comment]
Landed to align with master and feature/gltf will work on improved tests in separate PR.

**Pull request merged**